syntax = "proto3";

package market_intel_brain;

// Market Data Protocol Buffers Schema
// Optimized for high-frequency trading and real-time processing

// ===========================================
// CORE ENUMS
// ===========================================

enum DataSource {
  DATA_SOURCE_UNKNOWN = 0;
  DATA_SOURCE_BINANCE = 1;
  DATA_SOURCE_YAHOO_FINANCE = 2;
  DATA_SOURCE_FINNHUB = 3;
  DATA_SOURCE_ALPHA_VANTAGE = 4;
  DATA_SOURCE_NEWSAPI = 5;
  DATA_SOURCE_MARKETSTACK = 6;
  DATA_SOURCE_FMP = 7;
  DATA_SOURCE_POLYGON = 8;
  DATA_SOURCE_IEX = 9;
  DATA_SOURCE_QUANDL = 10;
  DATA_SOURCE_COINBASE = 11;
  DATA_SOURCE_KRAKEN = 12;
  DATA_SOURCE_BITFINEX = 13;
}

enum MarketType {
  MARKET_TYPE_UNKNOWN = 0;
  MARKET_TYPE_CRYPTO = 1;
  MARKET_TYPE_STOCK = 2;
  MARKET_TYPE_FOREX = 3;
  MARKET_TYPE_COMMODITY = 4;
  MARKET_TYPE_BOND = 5;
}

enum SentimentType {
  SENTIMENT_UNKNOWN = 0;
  SENTIMENT_POSITIVE = 1;
  SENTIMENT_NEGATIVE = 2;
  SENTIMENT_NEUTRAL = 3;
  SENTIMENT_MIXED = 4;
}

enum NewsCategory {
  NEWS_CATEGORY_UNKNOWN = 0;
  NEWS_CATEGORY_MARKET_NEWS = 1;
  NEWS_CATEGORY_COMPANY_NEWS = 2;
  NEWS_CATEGORY_ECONOMIC = 3;
  NEWS_CATEGORY_REGULATORY = 4;
  NEWS_CATEGORY_TECHNICAL = 5;
  NEWS_CATEGORY_ANALYSIS = 6;
}

// ===========================================
// CORE MESSAGES
// ===========================================

message MarketTick {
  // Core identification
  string symbol = 1;                    // Trading symbol (e.g., "BTCUSDT")
  DataSource source = 2;                // Data source
  MarketType market_type = 3;            // Market type
  
  // Price data
  double price = 4;                      // Current price
  double bid = 5;                        // Best bid price
  double ask = 6;                        // Best ask price
  double volume = 7;                     // 24h volume
  
  // Change data
  double change = 8;                     // Price change
  double change_percent = 9;               // Price change percentage
  double high_24h = 10;                 // 24h high
  double low_24h = 11;                  // 24h low
  
  // Timestamps
  string timestamp = 12;                   // ISO 8601 timestamp
  int64 timestamp_unix_ms = 13;            // Unix timestamp in milliseconds
  
  // Metadata
  string exchange = 14;                   // Exchange name
  string base_asset = 15;                 // Base asset (e.g., "BTC")
  string quote_asset = 16;                // Quote asset (e.g., "USDT")
  map<string, string> metadata = 17;       // Additional metadata
  
  // Quality indicators
  double confidence = 18;                  // Data confidence (0.0-1.0)
  int32 quality_score = 19;                // Quality score (0-100)
  bool is_delayed = 20;                   // Is data delayed
  int32 delay_ms = 21;                   // Delay in milliseconds
}

message NewsArticle {
  // Core identification
  string id = 1;                         // Unique article ID
  string title = 2;                       // Article title
  string content = 3;                      // Article content/summary
  string url = 4;                          // Article URL
  
  // Source information
  DataSource source = 5;                  // Data source
  string author = 6;                       // Author name
  string source_name = 7;                   // Source publication
  
  // Categorization
  NewsCategory category = 8;               // News category
  repeated string tags = 9;                  // Article tags
  repeated string symbols = 10;              // Related symbols
  
  // Timestamps
  string published_at = 11;                 // Publication timestamp
  int64 published_at_unix_ms = 12;          // Unix timestamp in milliseconds
  string created_at = 13;                   // Creation timestamp
  string updated_at = 14;                   // Last update timestamp
  
  // Content analysis
  string language = 15;                     // Content language
  int32 word_count = 16;                    // Word count
  double readability_score = 17;             // Readability score
  
  // Metadata
  map<string, string> metadata = 18;        // Additional metadata
  string image_url = 19;                    // Featured image URL
  repeated string related_articles = 20;       // Related article IDs
}

message SentimentData {
  // Core identification
  string symbol = 1;                       // Related symbol
  DataSource source = 2;                     // Data source
  
  // Sentiment analysis
  SentimentType sentiment = 3;               // Sentiment type
  double score = 4;                         // Sentiment score (-1.0 to 1.0)
  double confidence = 5;                      // Confidence score (0.0-1.0)
  
  // Analysis details
  string analysis_text = 6;                  // Analysis text
  repeated string keywords = 7;                // Key sentiment keywords
  double magnitude = 8;                      // Sentiment magnitude
  
  // Timestamps
  string timestamp = 9;                      // Analysis timestamp
  int64 timestamp_unix_ms = 10;               // Unix timestamp in milliseconds
  string data_window_start = 11;              // Data window start
  string data_window_end = 12;                // Data window end
  
  // Source data
  string source_text = 13;                    // Original source text
  repeated NewsArticle related_articles = 14;    // Related news articles
  
  // Metadata
  map<string, string> metadata = 15;          // Additional metadata
  string model_version = 16;                  // Sentiment model version
  int32 sample_size = 17;                     // Analysis sample size
}

message EconomicIndicator {
  // Core identification
  string indicator_name = 1;                 // Indicator name (e.g., "GDP", "CPI")
  string country = 2;                        // Country code
  DataSource source = 3;                      // Data source
  
  // Values
  double value = 4;                          // Current value
  double previous_value = 5;                   // Previous value
  double change = 6;                         // Change from previous
  double change_percent = 7;                   // Percentage change
  
  // Timestamps
  string timestamp = 8;                        // Release timestamp
  int64 timestamp_unix_ms = 9;                 // Unix timestamp in milliseconds
  string next_release = 10;                    // Next release date
  
  // Metadata
  string unit = 11;                          // Unit of measurement
  string frequency = 12;                      // Frequency (monthly, quarterly, etc.)
  map<string, string> metadata = 13;           // Additional metadata
  
  // Quality
  double confidence = 14;                     // Data confidence
  int32 revision_count = 15;                  // Number of revisions
}

// ===========================================
// BATCH MESSAGES
// ===========================================

message MarketDataBatch {
  // Batch identification
  string batch_id = 1;                       // Unique batch ID
  DataSource source = 2;                      // Data source
  
  // Batch content
  repeated MarketTick market_ticks = 3;        // Market data ticks
  repeated NewsArticle news_articles = 4;       // News articles
  repeated SentimentData sentiment_data = 5;     // Sentiment data
  repeated EconomicIndicator economic_indicators = 6; // Economic indicators
  
  // Batch metadata
  string timestamp = 7;                       // Batch creation timestamp
  int64 timestamp_unix_ms = 8;                 // Unix timestamp in milliseconds
  int32 total_items = 9;                      // Total items in batch
  
  // Processing information
  string processing_node = 10;                  // Processing node ID
  map<string, string> processing_metadata = 11; // Processing metadata
  int32 batch_size_bytes = 12;                 // Batch size in bytes
  double compression_ratio = 13;                // Compression ratio
}

message DataStream {
  // Stream identification
  string stream_id = 1;                       // Unique stream ID
  string stream_name = 2;                      // Stream name
  DataSource source = 3;                        // Data source
  
  // Stream configuration
  repeated string symbols = 4;                  // Subscribed symbols
  MarketType market_type = 5;                  // Market type
  map<string, string> filters = 6;             // Stream filters
  
  // Stream status
  bool is_active = 7;                         // Stream active status
  string last_activity = 8;                    // Last activity timestamp
  int64 messages_sent = 9;                    // Total messages sent
  int64 bytes_sent = 10;                       // Total bytes sent
  
  // Quality metrics
  double avg_latency_ms = 11;                  // Average latency
  double success_rate = 12;                    // Success rate
  int64 error_count = 13;                      // Error count
  
  // Timestamps
  string created_at = 14;                      // Stream creation timestamp
  string updated_at = 15;                      // Last update timestamp
}

// ===========================================
// CONTROL MESSAGES
// ===========================================

message SubscriptionRequest {
  // Subscription details
  string subscription_id = 1;                  // Unique subscription ID
  string client_id = 2;                        // Client identifier
  DataSource source = 3;                        // Data source
  
  // Data filters
  repeated string symbols = 4;                  // Symbols to subscribe
  MarketType market_type = 5;                  // Market type filter
  map<string, string> filters = 6;             // Additional filters
  
  // Subscription options
  bool real_time = 7;                          // Real-time updates
  bool include_history = 8;                     // Include historical data
  string history_start = 9;                     // History start timestamp
  string history_end = 10;                       // History end timestamp
  
  // Quality requirements
  double min_confidence = 11;                   // Minimum confidence
  int32 max_delay_ms = 12;                     // Maximum acceptable delay
  
  // Timestamps
  string created_at = 13;                      // Request timestamp
  int64 created_at_unix_ms = 14;                // Unix timestamp in milliseconds
}

message SubscriptionResponse {
  // Response details
  string subscription_id = 1;                  // Subscription ID
  bool success = 2;                            // Success status
  string error_message = 3;                     // Error message if failed
  
  // Subscription info
  repeated string symbols = 4;                  // Subscribed symbols
  MarketType market_type = 5;                  // Market type
  DataSource source = 6;                        // Data source
  
  // Stream information
  string stream_id = 7;                        // Stream ID
  string stream_endpoint = 8;                    // Stream endpoint
  
  // Quality metrics
  double expected_latency_ms = 9;                // Expected latency
  double success_rate = 10;                     // Expected success rate
  
  // Timestamps
  string timestamp = 11;                        // Response timestamp
  int64 timestamp_unix_ms = 12;                 // Unix timestamp in milliseconds
}

message Acknowledgment {
  // Acknowledgment details
  string message_id = 1;                       // Original message ID
  string subscription_id = 2;                    // Subscription ID
  string client_id = 3;                          // Client ID
  
  // Acknowledgment status
  bool success = 4;                             // Success status
  string error_code = 5;                         // Error code if failed
  string error_message = 6;                      // Error message if failed
  
  // Processing information
  string processed_by = 7;                       // Processing node
  string processing_time = 8;                    // Processing timestamp
  int64 processing_time_unix_ms = 9;             // Unix timestamp in milliseconds
  double processing_latency_ms = 10;              // Processing latency
  
  // Quality metrics
  int32 queue_position = 11;                     // Position in queue
  int32 retry_count = 12;                        // Number of retries
  double success_rate = 13;                      // Success rate for this client
  
  // Metadata
  map<string, string> metadata = 14;              // Additional metadata
  string next_retry_time = 15;                   // Next retry timestamp
}

// ===========================================
// HEALTH AND MONITORING
// ===========================================

message HealthCheck {
  // Health status
  bool healthy = 1;                             // Overall health status
  string status = 2;                             // Status description
  string version = 3;                             // Service version
  
  // Component status
  map<string, bool> components = 4;               // Component health status
  repeated string failing_components = 5;            // Failing components
  
  // Performance metrics
  double uptime_percentage = 6;                   // Uptime percentage
  int64 total_requests = 7;                       // Total requests handled
  int64 successful_requests = 8;                  // Successful requests
  int64 failed_requests = 9;                      // Failed requests
  double success_rate = 10;                       // Success rate
  
  // Resource usage
  double cpu_usage = 11;                         // CPU usage percentage
  double memory_usage = 12;                       // Memory usage percentage
  double disk_usage = 13;                         // Disk usage percentage
  int64 active_connections = 14;                  // Active connections
  
  // Timestamps
  string timestamp = 15;                           // Check timestamp
  int64 timestamp_unix_ms = 16;                    // Unix timestamp in milliseconds
  string last_restart = 17;                        // Last restart timestamp
}

message PerformanceMetrics {
  // Metrics identification
  string node_id = 1;                            // Node identifier
  string service_name = 2;                        // Service name
  string timestamp = 3;                           // Metrics timestamp
  int64 timestamp_unix_ms = 4;                     // Unix timestamp in milliseconds
  
  // Throughput metrics
  double messages_per_second = 5;                  // Messages per second
  double bytes_per_second = 6;                     // Bytes per second
  int64 total_messages = 7;                        // Total messages
  int64 total_bytes = 8;                           // Total bytes
  
  // Latency metrics
  double avg_latency_ms = 9;                      // Average latency
  double p50_latency_ms = 10;                     // 50th percentile latency
  double p95_latency_ms = 11;                     // 95th percentile latency
  double p99_latency_ms = 12;                     // 99th percentile latency
  double max_latency_ms = 13;                     // Maximum latency
  
  // Error metrics
  double error_rate = 14;                         // Error rate
  int64 total_errors = 15;                         // Total errors
  map<string, int64> error_counts = 16;            // Error counts by type
  
  // Resource metrics
  double cpu_usage = 17;                           // CPU usage percentage
  double memory_usage = 18;                         // Memory usage percentage
  int64 active_connections = 19;                    // Active connections
  int64 queue_size = 20;                           // Queue size
  
  // Quality metrics
  double success_rate = 21;                         // Success rate
  double data_quality_score = 22;                   // Data quality score
  int64 duplicate_detections = 23;                  // Duplicate detections
  int64 late_messages = 24;                        // Late messages
  
  // Custom metrics
  map<string, double> custom_metrics = 25;           // Custom metrics
  map<string, string> metadata = 26;                // Additional metadata
}

// ===========================================
// ERROR HANDLING
// ===========================================

message ErrorInfo {
  // Error details
  string error_id = 1;                            // Unique error ID
  string error_code = 2;                          // Error code
  string error_message = 3;                        // Error message
  string error_type = 4;                           // Error type
  
  // Context information
  string service_name = 5;                        // Service name
  string operation = 6;                            // Operation being performed
  string client_id = 7;                            // Client ID
  string request_id = 8;                           // Request ID
  
  // Error details
  map<string, string> context = 9;                // Error context
  repeated string stack_trace = 10;                 // Stack trace
  string severity = 11;                            // Error severity
  
  // Timestamps
  string timestamp = 12;                           // Error timestamp
  int64 timestamp_unix_ms = 13;                    // Unix timestamp in milliseconds
  string resolved_at = 14;                         // Resolution timestamp
  bool is_resolved = 15;                           // Resolution status
  
  // Metadata
  map<string, string> metadata = 16;                // Additional metadata
  string related_error_id = 17;                     // Related error ID
  int32 retry_count = 18;                          // Number of retries
}
